File: src/seemps/state/mpssum.py
from __future__ import annotations
import numpy as np
from math import sqrt
from collections.abc import Iterable, Sequence
from typing import cast
from ..tools import InvalidOperation
from ..typing import Weight, Vector, Tensor3
from .environments import scprod


class MPSSum:
    """Class representing a weighted sum (or difference) of two or more :class:`MPS`.

    This class is an intermediate representation for the linear combination of
    MPS quantum states. Assume that :math:`\\psi, \\phi` and :math:`\\xi` are
    MPS and :math:`a, b, c` some real or complex numbers. The addition
    :math:`a \\psi - b \\phi + c \\xi` can be stored as
    `MPSSum([a, -b, c], [ψ, ϕ, ξ])`.


    Parameters
    ----------
    weights : list[Weight]
        Real or complex numbers representing the weights of the linear combination.
    states : list[MPS]
        List of matrix product states weighted.
    """

    weights: list[Weight]
    states: list[MPS]
    size: int

    #
    # This class contains all the matrices and vectors that form
    # a Matrix-Product State.
    #
    __array_priority__: int = 10000

    def __init__(
        self,
        weights: Iterable[Weight],
        states: Iterable[MPS | MPSSum],
        check_args: bool = True,
    ):
        if check_args:
            self.weights = new_weights = []
            self.states = new_states = []
            for w, s in zip(weights, states):
                if isinstance(s, MPS):
                    new_weights.append(w)
                    new_states.append(s)
                elif isinstance(s, MPSSum):
                    new_weights.extend(w * wi for wi in s.weights)
                    new_states.extend(s.states)
                else:
                    raise ValueError(s)
            self.size = new_states[0].size
        else:
            self.weights = list(weights)
            self.states = cast(list[MPS], list(states))
            self.size = self.states[0].size

    def as_mps(self) -> MPS:
        return self.join()

    def copy(self) -> MPSSum:
        """Return a shallow copy of the MPS sum and its data. Does not copy
        the states, only the list that stores them."""
        return MPSSum(self.weights.copy(), self.states.copy())

    def __copy__(self) -> MPSSum:
        return self.copy()

    def __add__(self, state: MPS | MPSSum) -> MPSSum:
        """Add `self + state`, incorporating it to the lists."""
        match state:
            case MPS():
                return MPSSum(
                    self.weights + [1.0], self.states + [state], check_args=False
                )
            case MPSSum(weights=w, states=s):
                return MPSSum(self.weights + w, self.states + s, check_args=False)
            case _:
                raise InvalidOperation("+", self, state)

    def __sub__(self, state: MPS | MPSSum) -> MPSSum:
        """Subtract `self - state`, incorporating it to the lists."""
        match state:
            case MPS():
                return MPSSum(
                    self.weights + [-1], self.states + [state], check_args=False
                )
            case MPSSum(weights=w, states=s):
                return MPSSum(
                    self.weights + [-wi for wi in w], self.states + s, check_args=False
                )
            case _:
                raise InvalidOperation("-", self, state)

    def __mul__(self, n: Weight) -> MPSSum:
        """Rescale the linear combination `n * self` for scalar `n`."""
        if isinstance(n, (int, float, complex)):
            return MPSSum([n * w for w in self.weights], self.states, check_args=False)
        raise InvalidOperation("*", self, n)

    def __rmul__(self, n: Weight) -> MPSSum:
        """Rescale the linear combination `self * n` for scalar `n`."""
        if isinstance(n, (int, float, complex)):
            return MPSSum([n * w for w in self.weights], self.states, check_args=False)
        raise InvalidOperation("*", n, self)

    def to_vector(self) -> Vector:
        """Return the wavefunction of this quantum state."""
        return sum(wa * A.to_vector() for wa, A in zip(self.weights, self.states))  # type: ignore # pyright: ignore[reportReturnType]

    def _joined_tensors(self, i: int, L: int) -> Tensor3:
        """Join the tensors from all MPS into bigger tensors."""
        As: list[Tensor3] = [s[i] for s in self.states]
        if i == 0:
            return np.concatenate([w * A for w, A in zip(self.weights, As)], axis=2)
        if i == L - 1:
            return np.concatenate(As, axis=0)

        DL: int = 0
        DR: int = 0
        d: int = 0
        w: Weight = 0
        for A in As:
            a, d, b = A.shape
            DL += a
            DR += b
            w += A[0, 0, 0]
        B = np.zeros((DL, d, DR), dtype=type(w))
        DL = 0
        DR = 0
        for A in As:
            a, d, b = A.shape
            B[DL : DL + a, :, DR : DR + b] = A
            DL += a
            DR += b
        return B

    def join(self) -> MPS:
        """Create an `MPS` by combining all tensors from all states in
        the linear combination.

        Returns
        -------
        MPS
            Quantum state approximating this sum.
        """
        L = self.size
        return MPS([self._joined_tensors(i, L) for i in range(L)])

    def join_canonical(self, *args, **kwdargs) -> CanonicalMPS:  # pyright: ignore[reportMissingParameterType]
        """Similar to join() but return canonical form"""
        return CanonicalMPS(self.join(), *args, **kwdargs)

    def conj(self) -> MPSSum:
        """Return the complex-conjugate of this quantum state."""
        return MPSSum(
            [np.conj(w) for w in self.weights], [state.conj() for state in self.states]
        )

    def norm_squared(self) -> float:
        """Norm-2 squared :math:`\\Vert{\\psi}\\Vert^2` of this MPS."""
        w = self.weights
        s = self.states
        L = len(w)
        return abs(
            sum(
                (w[i].conjugate() * w[j] * scprod(s[i], s[j])).real
                * (1 if i == j else 2)
                for i in range(L)
                for j in range(i, L)
            )
        )

    def norm(self) -> float:
        """Norm-2 :math:`\\Vert{\\psi}\\Vert^2` of this MPS."""
        return sqrt(self.norm_squared())

    def error(self) -> float:
        """Upper bound of the norm-2 error accumulated in this MPS."""
        return sum(
            abs(weight) * state._error
            for weight, state in zip(self.weights, self.states)
        )

    def delete_zero_components(self) -> float:
        """Compute the norm-squared of the linear combination of weights and
        states and eliminate states that are zero or have zero weight."""
        c: float = 0.0
        final_weights: list[Weight] = []
        final_states: list[MPS] = []
        for wi, si in zip(self.weights, self.states):
            wic = wi.conjugate()
            ni = (wic * wi).real * si.norm_squared()
            if ni:
                for wj, sj in zip(final_weights, final_states):
                    c += 2 * (wic * wj * scprod(si, sj)).real
                final_states.append(si)
                final_weights.append(wi)
                c += ni
        L = len(final_weights)
        if L < len(self.states):
            if L == 0:
                self.weights = [0.0]
                self.states = [self.states[0].zero_state()]
                return 0.0
            else:
                self.weights = final_weights
                self.states = final_states
        return abs(c)

    # TODO: We have to change the signature and working of this function, so that
    # 'sites' only contains the locations of the _new_ sites, and 'L' is no longer
    # needed. In this case, 'dimensions' will only list the dimensions of the added
    # sites, not all of them.
    def extend(
        self,
        L: int,
        sites: Sequence[int] | None = None,
        dimensions: int | list[int] = 2,
        state: Vector | None = None,
    ) -> MPSSum:
        return MPSSum(
            self.weights,
            [A.extend(L, sites, dimensions, state) for A in self.states],
            check_args=False,
        )

    def reverse(self) -> MPSSum:
        """Reverse the sites (see :meth:`~seemps.state.MPS.reverse`)."""
        return MPSSum(
            self.weights[-1::],
            [state.reverse() for state in reversed(self.states)],
        )


from .canonical_mps import CanonicalMPS  # noqa: E402
from .mps import MPS  # noqa: E402


def to_mps(mps_or_sum: MPS | MPSSum) -> MPS:
    if isinstance(mps_or_sum, MPSSum):
        return mps_or_sum.join()
    return mps_or_sum

File: src/seemps/state/mpssum.cc
#include "mps.h"

namespace seemps {

MPSSum::MPSSum(py::object weights, py::object states, bool check_args) {
  if (!check_args) {
    weights_ = py::list(weights);
    mps_ = py::list(states);
  } else {
    auto L = py::len(weights);
    if (L != py::len(states)) {
      throw std::invalid_argument(
          "Lists of weights does not match list of states in MPSSum");
    }
    if (L == 0) {
      throw std::invalid_argument("MPSSum requires a non-empty list of states");
    }
    weights_ = py::list();
    mps_ = py::list();
    for (int i = 0; i < L; ++i) {
      auto wi = weights[i];
      auto si = states[i];
      if (py::isinstance<MPSSum>(si)) {
        append(wi, py::cast<MPSSum>(si));
      } else if (py::isinstance<MPS>(si)) {
        append(wi, si);
      } else {
        throw std::invalid_argument(
            "MPSSum argument did not contain a valid MPS");
      }
    }
  }
  size_ = mps(0).size();
}

MPSSum::MPSSum(const MPS &mps)
    : weights_{py::empty_list(1)}, mps_{py::empty_list(1)}, size_{mps.size()} {
  weights_[0] = 1.0;
  mps_[0] = py::cast(mps);
}

MPSSum MPSSum::copy() const {
  return MPSSum(py::copy(weights_), py::copy(mps_), false);
}

MPSSum MPSSum::deepcopy() const {
  py::list mps_list;
  std::for_each(py::begin(mps_), py::end(mps_), [&](py::object a) -> void {
    py::object mps_copy;
    if (py::isinstance<MPS>(a)) {
      mps_copy = py::cast(py::cast<MPS &>(a).deepcopy());
    } else {
      mps_copy = py::cast(py::cast<CanonicalMPS &>(a).deepcopy());
    }
    mps_list.append(mps_copy);
  });
  return MPSSum(py::copy(weights_), mps_list, false);
}

MPSSum MPSSum::conj() const {
  auto output = copy();
  for (size_t i = 0; i < sum_size(); ++i) {
    output.weights_[i] = py::conj(output.weights_[i]);
    py::cast<MPS &>(output.mps_[i]).conj_in_place();
  }
  return output;
}

double MPSSum::norm_squared() const {
  double output = 0.0;
  for (size_t i = 0; i < sum_size(); ++i) {
    auto wi_conj = std::conj(py::cast<std::complex<double>>(weight(i)));
    const auto &si = mps(i);
    for (size_t j = i; j < sum_size(); ++j) {
      auto wj = py::cast<std::complex<double>>(weight(j));
      auto sij = py::cast<std::complex<double>>(scprod(si, mps(j)));
      auto s = (wi_conj * wj * sij).real() * ((i == j) ? 1 : 2);
      output += s;
    }
  }
  return std::abs(output);
}

double MPSSum::error() const {
  double output = 0.0;
  for (size_t i = 0; i < sum_size(); ++i) {
    auto wi = py::cast<std::complex<double>>(weights_[i]);
    auto error = py::cast<const MPS &>(mps_[i]).error();
    output += std::abs(wi) * error;
  }
  return output;
}

Weight MPSSum::weight(int site) const { return weights_[site]; }

MPS &MPSSum::mps(int site) const { return py::cast<MPS &>(mps_[site]); }

double MPSSum::norm() const { return std::sqrt(norm_squared()); }

void MPSSum::append(const Weight &weight, const py::object &mps) {
  weights_.append(weight);
  mps_.append(mps);
}

void MPSSum::append(const Weight &factor, const MPSSum &states) {
  const auto &weights = states.weights_;
  const auto &mps = states.mps_;
  for (size_t i = 0; i < states.sum_size(); ++i) {
    append(factor * weights[i], mps[i]);
  }
}

static double state_norm_squared(const py::object &state) {
  if (py::isinstance<CanonicalMPS>(state)) {
    return py::cast<const CanonicalMPS &>(state).norm_squared();
  } else {
    return py::cast<const MPS &>(state).norm_squared();
  }
}

double MPSSum::delete_zero_components() {
  double c = 0.0;
  py::list final_weights;
  py::list final_states;
  for (int i = 0; i < sum_size(); ++i) {
    std::complex<double> wi = py::cast<std::complex<double>>(weights_[i]);
    if (wi.real() != 0.0 || wi.imag() != 0) {
      auto wic = std::conj(wi);
      auto statei = mps_[i];
      const auto &si = py::cast<const MPS &>(statei);
      auto ni = (wic * wi).real() * state_norm_squared(statei);
      if (ni) {
        for (int j = 0; j < py::len(final_weights); ++j) {
          auto wj = py::cast<std::complex<double>>(final_weights[j]);
          auto sj = py::cast<const MPS &>(final_states[j]);
          c += 2 * (wic * wj * py::cast<std::complex<double>>(scprod(si, sj)))
                       .real();
        }
        final_weights.append(weights_[i]);
        final_states.append(statei);
        c += ni;
      }
    }
  }
  auto L = final_weights.size();
  if (L == 0) {
    final_weights.append(py::cast(0.0));
    final_states.append(py::cast<const MPS &>(mps_[0]).zero_state());
    weights_ = std::move(final_weights);
    mps_ = std::move(final_states);
    return 0.0;
  }
  if (L < sum_size()) {
    weights_ = std::move(final_weights);
    mps_ = std::move(final_states);
  }
  return std::abs(c);
}

} // namespace seemps
